ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/LfValueTranslation.scala

```diff
@@ -0,0 +1,222 @@
+
+  private def serializeNullableKeyOrThrow(c: Create): Option[Array[Byte]] =
+    c.key.map(
+      k =>
+        ValueSerializer.serializeValue(
+          value = k.key,
+          errorContext = cantSerialize(attribute = "key", forContract = c.coid),
+      )
+    )
+
+  private def serializeExerciseArgOrThrow(e: Exercise): Array[Byte] =
+    ValueSerializer.serializeValue(
+      value = e.chosenValue,
+      errorContext = cantSerialize(attribute = "exercise argument", forContract = e.targetCoid),
+    )
+
+  private def serializeNullableExerciseResultOrThrow(e: Exercise): Option[Array[Byte]] =
+    e.exerciseResult.map(
+      exerciseResult =>
+        ValueSerializer.serializeValue(
+          value = exerciseResult,
+          errorContext = cantSerialize(attribute = "exercise result", forContract = e.targetCoid),
+      )
+    )
+
+  // Doesn't go through caching, for now caching is limited to events
+  def serialize(contractId: ContractId, createArgument: LfValue): Vector[NamedParameter] =
+    Vector[NamedParameter](
+      "create_argument" -> serializeCreateArgOrThrow(contractId, createArgument),
+    )
+
+  def serialize(eventId: EventId, create: Create): Vector[NamedParameter] = {
+    cache.put(
+      key = LfValueTranslation.Cache.Key(eventId),
+      value = LfValueTranslation.Cache.Value.Create(create.coinst.arg, create.key.map(_.key)),
+    )
+    Vector[NamedParameter](
+      "create_argument" -> serializeCreateArgOrThrow(create),
+      "create_key_value" -> serializeNullableKeyOrThrow(create),
+    )
+  }
+
+  def serialize(eventId: EventId, exercise: Exercise): Vector[NamedParameter] = {
+    cache.put(
+      key = LfValueTranslation.Cache.Key(eventId),
+      value = LfValueTranslation.Cache.Value.Exercise(exercise.chosenValue, exercise.exerciseResult),
+    )
+    Vector[NamedParameter](
+      "exercise_argument" -> serializeExerciseArgOrThrow(exercise),
+      "exercise_result" -> serializeNullableExerciseResultOrThrow(exercise),
+    )
+  }
+
+  private def toApiValue(
+      value: LfValue,
+      verbose: Boolean,
+      attribute: => String,
+  ): ApiValue =
+    LfEngineToApi.assertOrRuntimeEx(
+      failureContext = s"attempting to deserialize persisted $attribute to value",
+      LfEngineToApi
+        .lfVersionedValueToApiValue(
+          verbose = verbose,
+          value = value,
+        ),
+    )
+
+  private def toApiRecord(
+      value: LfValue,
+      verbose: Boolean,
+      attribute: => String,
+  ): ApiRecord =
+    LfEngineToApi.assertOrRuntimeEx(
+      failureContext = s"attempting to deserialize persisted $attribute to record",
+      LfEngineToApi
+        .lfVersionedValueToApiRecord(
+          verbose = verbose,
+          recordValue = value,
+        ),
+    )
+
+  def deserialize[E](raw: Raw.Created[E], verbose: Boolean): CreatedEvent = {
+    val key = LfValueTranslation.Cache.Key(raw.partial.eventId)
+    val create =
+      cache
+        .getIfPresent(key = key)
```

Comment:
Why not just use `.get(key, _ => …)`?

Reply:
Because I don't want to recompute most recently queried events, only the most recently indexed ones.